### PR TITLE
Bl/#2373 actor that collects charging events

### PR DESCRIPTION
**Pull request by bhavyalatha26**
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/2376)

### CODE REVIEW

**Comment by JustinPihony** [comment]
Test!

File: src/main/scala/beam/agentsim/agents/vehicles/ChargingEventsAccumulator.scala [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      chargingEventsBuffer += e
+
+    case Finish =>
+      informExternalSystem(chargingEventsBuffer)
+      context.stop(self)
+
+  }
+
+  private def informExternalSystem(chargingEventsBuffer: ListBuffer[org.matsim.api.core.v01.events.Event]): Unit = {
+    //TODO implement this stub later
```
Comment (JustinPihony) [inline src/main/scala/beam/agentsim/agents/vehicles/ChargingEventsAccumulator.scala]: Is this still a draft PR? Should this be implemented?

File: src/main/resources/beam-template.conf [hunk 2]
```diff
@@ -218,8 +218,10 @@ beam.agentsim.agents.vehicles.sharedFleets = [
     }
   }
 ]
-beam.agentsim.agents.vehicles.collectChargingEvents=true
-beam.agentsim.agents.vehicles.collectChargingEventsIntervalInSeconds=300
+
+beam.agentsim.agents.vehicles.collectEvents=true
```
Comment (wrashid) [inline src/main/resources/beam-template.conf]: default should be false

File: test/input/beamville/beam.conf [hunk 3]
```diff
@@ -47,8 +47,10 @@ beam.agentsim.agents.vehicles.fuelTypesFilePath = ${beam.inputDirectory}"/beamFu
 beam.agentsim.agents.vehicles.vehicleTypesFilePath = ${beam.inputDirectory}"/vehicleTypes.csv"
 beam.agentsim.agents.vehicles.vehiclesFilePath = ${beam.inputDirectory}"/vehicles.csv"
 beam.agentsim.agents.vehicles.sharedFleets = []
-beam.agentsim.agents.vehicles.collectChargingEvents=true
-beam.agentsim.agents.vehicles.collectChargingEventsIntervalInSeconds=300
+
+beam.agentsim.agents.vehicles.collectEvents=true
+beam.agentsim.agents.vehicles.collectEventsIntervalInSeconds=300
+beam.agentsim.agents.vehicles.collectEventTypes="charging"
```
Comment (wrashid) [inline test/input/beamville/beam.conf]: where is this used in the code?

File: src/main/scala/beam/sim/LoggingEventsManager.scala [hunk 4]
```diff
@@ -34,18 +34,18 @@ class LoggingEventsManager @Inject()(config: Config) extends EventsManager with
...
   }
 
   override def processEvent(event: Event): Unit = {
-    chargingEventsAccumulator match {
+    eventsAccumulator match {
       case Some(accumulator) =>
         event match {
           case e @ (_: ChargingPlugInEvent | _: ChargingPlugOutEvent | _: RefuelSessionEvent) =>
-            accumulator ! e
+            accumulator ! EventsAccumulator.ProcessChargingEvents(e)
```
Comment (wrashid) [inline src/main/scala/beam/sim/LoggingEventsManager.scala]: need to make this generic

File: src/main/scala/beam/agentsim/agents/vehicles/ChargingEventsAccumulator.scala (hunk 1 above)
Comment (bhavyalatha26) [reply src/main/scala/beam/agentsim/agents/vehicles/ChargingEventsAccumulator.scala]: No this stub will have no implementation for now. Ya, the PR is still a draft one

File: src/main/resources/beam-template.conf (hunk 2 above)
Comment (bhavyalatha26) [reply src/main/resources/beam-template.conf]: ok

Comment (wrashid) [reply src/main/resources/beam-template.conf]: use beam.agentsim.collectEvents

File: src/main/resources/beam-template.conf [hunk 5]
```diff
@@ -218,8 +218,10 @@ beam.agentsim.agents.vehicles.sharedFleets = [
...
+beam.agentsim.agents.vehicles.collectEventsIntervalInSeconds=300
+beam.agentsim.agents.vehicles.collectEventTypes="charging"
```
Comment (wrashid) [inline src/main/resources/beam-template.conf]: let us remove collectEventTypes for now

**Comment by bhavyalatha26** [comment]
Test!

File: src/main/resources/beam-template.conf (hunk 5 above)
Comment (wrashid) [inline src/main/resources/beam-template.conf]: remove this line

**Comment by haitamlaarabi** [comment]
@wrashid @bhavyalatha26 
I don't see the reason of collecting events so to be transmitted as messages to an Actor. Why wouldn't be the eventsAcumulator both an Actor an EventHandler itself, without need to transmit messages ? 
+ bonus point: better modularity

**Comment by wrashid** [comment]
> @wrashid @bhavyalatha26
> I don't see the reason of collecting events so to be transmitted as messages to an Actor. Why wouldn't be the eventsAcumulator both an Actor an EventHandler itself, without need to transmit messages ?
> 
> * bonus point: better modularity

EventHandlers are running single threaded and some are writing outputs. We do not want to slow down our main simulation, by making it dependent on them (e.g. see that the agentsim scheduler is awaiting trigger release). For the same reason we are sending the events to the accumulator actor in the first line of processEvent to avoid any overheads which might be cause by event handlers.

**Pull request merged**